File: codility/lessons/1 - iterations/binary gap/binary_gap_procedural.js
// Source: https://app.codility.com/programmers/lessons/1-iterations/binary_gap/

function solution(number) {
    const BINARY = convertNumberToBinaryString(number);
    const LONGEST_BINARY_GAP = findLongestBinaryGap(BINARY);
    return LONGEST_BINARY_GAP;
}

    function convertNumberToBinaryString(number) {
        return (number >>> 0).toString(2);
    }

    function findLongestBinaryGap(binary) {
        let longestBinaryGap = 0;

        for (let index = {'i': 0, 'j': 0, 'k': 0}; Object.values(index).includes(-1) === false;) {
            const GAP_LENGTH = findGapLength(binary, index);
            longestBinaryGap = Math.max(GAP_LENGTH, longestBinaryGap);
        }

        return longestBinaryGap;
    }

        function findGapLength(binary, index) {
            index.i = binary.indexOf('1', index.j);

            if (index.i != -1) {
                index.j = binary.indexOf('0', index.i + 1);

                if (index.j != -1) {
                    index.k = binary.indexOf('1', index.j + 1);

                    if (index.k != -1) {
                        const GAP_LENGTH = index.k - index.j;
                        return GAP_LENGTH;
                    }
                    else
                        return 0;
                }
                else
                    return 0;
            }
            else
                return 0;
        }

File: codility/lessons/1 - iterations/binary gap/binary_gap_procedural.cs
// Source: https://app.codility.com/programmers/lessons/1-iterations/binary_gap/

using System;
using System.Collections.Generic;

class Solution
{
    public int solution(int number)
    {
        string binary = _convertIntToBinaryString(number);
        int longestBinaryGap = _findLongestBinaryGap(ref binary);
        return longestBinaryGap;
    }

        private string _convertIntToBinaryString(int number)
        {
            return Convert.ToString(number, 2);
        }

        private int _findLongestBinaryGap(ref string binary)
        {
            int longestBinaryGap = 0;

            for (List<int> index = new List<int>(new int[3]); !index.Contains(-1);)
            {
                int gapLength = _findGapLength(ref binary, ref index);
                longestBinaryGap = Math.Max(gapLength, longestBinaryGap);
            }

            return longestBinaryGap;
        }

            private int _findGapLength(ref string binary, ref List<int> index)
            {
                index[0] = binary.IndexOf('1', index[1]);
                if (index[0] != -1)
                {
                    index[1] = binary.IndexOf('0', index[0] + 1);
                    if (index[1] != -1)
                    {
                        index[2] = binary.IndexOf('1', index[1] + 1);
                        if (index[2] != -1)
                        {
                            int gapLength = index[2] - index[1];
                            return gapLength;
                        }
                        else
                            return 0;
                    }
                    else
                        return 0;
                }
                else
                    return 0;
            }
}
